api/envoy/config/core/v3/protocol.proto

```diff
@@ -417,6 +417,9 @@ message Http1ProtocolOptions {
   // <envoy_v3_api_field_extensions.http.header_validators.envoy_default.v3.HeaderValidatorConfig.restrict_http_methods>`
   // to reject custom methods.
   bool allow_custom_methods = 10 [(xds.annotations.v3.field_status).work_in_progress = true];
+
+  // Ignore HTTP/1.1 upgrade headers rather than default rejection
+  bool ignore_http_11_upgrade = 11;
```

Comment:
From the naming, the options will ignore all upgrade. But actually the implementation didn't works at that way.

Reply:
ignore_http_11_tls_upgrade ?

Reply:
or repeated string of upgrade to ignore, so it can be extended without adding multiple config fields.

I'd strongly prefer if we're going to ignore we do what we do for h2c and remove the unsupported upgrade.  If Envoy is not going to treat an upgrade as an upgrade, but an upstream is, I think you run into  problems.  If we want to handle the case that a client asks for an upgrade which neither Envoy or the upstream handle, removing the upgrade header is a clear signal that we're handling the upgrade.

If you have a reason to pass the upgrade header through, this PR should not land until you also add handling verifying that the upstream did not attempt to actually upgrade
/wait-any

Reply:
Forwarding the `Upgrade` header is not valid by my reading of the spec:

>    When Upgrade is sent, the sender MUST also send a Connection header
>    field ([Section 6.1](https://datatracker.ietf.org/doc/html/rfc7230#section-6.1)) that contains an "upgrade" connection option, in
>    order to prevent Upgrade from being accidentally forwarded by
>    intermediaries that might not implement the listed protocols.

`Connection` header is hop-by-hop so it cannot be forwarded. `Upgrade` cannot be sent without a `Connection` header. Therefore `Upgrade` cannot be forwarded (or at the very least is extraneous).

Reply:
perfect, so a repeated string of upgrade headers to clear makes the most sense to me

Reply:
I think having an (additional?) way to ignore all upgrades may be desired. For example, we want to configure to ignore all upgrades because users may set completely arbitrary ones and we cannot know them all.